src/core/models.py

```diff
@@ -5,51 +5,86 @@
 """Collection of state objects for the Kafka relations, apps and units."""
 
 import logging
-from typing import MutableMapping
 
+from charms.data_platform_libs.v0.data_interfaces import Data, DataPeerData, DataPeerUnitData
 from charms.zookeeper.v0.client import QuorumLeaderNotFoundError, ZooKeeperManager
-from kazoo.exceptions import AuthFailedError, NoNodeError
+from kazoo.client import AuthFailedError, NoNodeError
 from ops.model import Application, Relation, Unit
-from tenacity import retry, retry_if_not_result, stop_after_attempt, wait_fixed
+from tenacity import retry
+from tenacity.retry import retry_if_not_result
+from tenacity.stop import stop_after_attempt
+from tenacity.wait import wait_fixed
+from typing_extensions import override
 
-from literals import INTERNAL_USERS, Substrate
+from literals import INTERNAL_USERS, SECRETS_APP, Substrates
 
 logger = logging.getLogger(__name__)
 
 
-class StateBase:
-    """Base state object."""
+class RelationState:
+    """Relation state object."""
 
     def __init__(
-        self, relation: Relation | None, component: Unit | Application, substrate: Substrate
+        self,
+        relation: Relation | None,
+        data_interface: Data,
+        component: Unit | Application | None,
+        substrate: Substrates,
     ):
         self.relation = relation
+        self.data_interface = data_interface
         self.component = component
         self.substrate = substrate
+        self.relation_data = self.data_interface.as_dict(self.relation.id) if self.relation else {}
 
-    @property
-    def relation_data(self) -> MutableMapping[str, str]:
-        """The raw relation data."""
-        if not self.relation:
-            return {}
-
-        return self.relation.data[self.component]
+    def __bool__(self) -> bool:
+        """Boolean evaluation based on the existence of self.relation."""
+        try:
+            return bool(self.relation)
+        except AttributeError:
+            return False
 
     def update(self, items: dict[str, str]) -> None:
         """Writes to relation_data."""
-        if not self.relation:
-            return
+        delete_fields = [key for key in items if not items[key]]
+        update_content = {k: items[k] for k in items if k not in delete_fields}
+
+        self.relation_data.update(update_content)
 
-        self.relation_data.update(items)
+        for field in delete_fields:
+            del self.relation_data[field]
```

Comment:
Certainly my hope is that we expose this `RelationState` object in the lib, so that it's shared across the team.
That said, I think it needs more time and usage from us to propose the solution, if it works, what needs to change etc.